subprojects/launcher/src/main/java/org/gradle/launcher/bootstrap/MixInCoreTypesTransformingClassLoader.java

```diff
@@ -0,0 +1,231 @@
+        return classWriter.toByteArray();
+    }
+
+    /**
+     * Adds overrides with internal return types for {@link org.gradle.api.Task#getInputs()} and
+     * {@link org.gradle.api.Task#getOutputs()} to {@link org.gradle.api.internal.TaskInternal}.
+     */
+    private static class TaskInternalTransformer extends ClassVisitor {
+        public TaskInternalTransformer(ClassVisitor cv) {
+            super(Opcodes.ASM5, cv);
+        }
+
+        @Override
+        public void visitEnd() {
+            addInterfaceMethod(cv,
+                getType("org.gradle.api.internal.TaskInputsInternal"),
+                "getInputs"
+            );
+            addInterfaceMethod(cv,
+                getType("org.gradle.api.internal.TaskOutputsInternal"),
+                "getOutputs"
+            );
+            super.visitEnd();
+        }
+    }
+
+    /**
+     * Adds overrides with internal return types for {@link org.gradle.api.Task#getInputs()} and
+     * {@link org.gradle.api.Task#getOutputs()} to {@link org.gradle.api.internal.AbstractTask}.
+     */
+    private static class AbstractTaskTransformer extends ClassVisitor {
+        public AbstractTaskTransformer(ClassVisitor cv) {
+            super(Opcodes.ASM5, cv);
+        }
+
+        @Override
+        public void visitEnd() {
+            addBridgeMethod(cv,
+                getType("org.gradle.api.internal.TaskInputsInternal"),
+                "getInputs",
+                getType("org.gradle.api.Task"),
+                getType("org.gradle.api.tasks.TaskInputs")
+            );
+            addBridgeMethod(cv,
+                getType("org.gradle.api.internal.TaskOutputsInternal"),
+                "getOutputs",
+                getType("org.gradle.api.Task"),
+                getType("org.gradle.api.tasks.TaskOutputs")
+            );
+            super.visitEnd();
+        }
+    }
+
+    /**
+     * Adds overrides with internal return types for {@link org.gradle.api.tasks.TaskInputs#file()} and
+     * {@code dir()} and {@code files()} to {@link org.gradle.api.internal.tasks.DefaultTaskInputs}.
+     */
+    private static class TaskInputsInternalTransformer extends ClassVisitor {
+        public TaskInputsInternalTransformer(ClassVisitor cv) {
+            super(Opcodes.ASM5, cv);
+        }
+
+        @Override
+        public void visitEnd() {
+            addInterfaceMethod(cv,
+                getType("org.gradle.api.internal.tasks.TaskInputFilePropertyBuilderInternal"),
+                "file",
+                Type.getType(Object.class)
+            );
+            addInterfaceMethod(cv,
+                getType("org.gradle.api.internal.tasks.TaskInputFilePropertyBuilderInternal"),
+                "dir",
+                Type.getType(Object.class));
+            addInterfaceMethod(cv,
+                getType("org.gradle.api.internal.tasks.TaskInputFilePropertyBuilderInternal"),
+                "files",
+                Type.getType(Object[].class));
```

Comment:
We could, but this method will never be compiled against, so I don't think it's necessary. WDYT?